### ISSUE TITLE
Support for registering callback handler when fiber is suspended and resumed

### ISSUE THREAD

**Issue by nickman** [puniverse/quasar#191]
In order to better integrate with third party libraries that retain thread specific references, it would be helpful to provide a callback before the park and after the unpark of a fiber. These items are opinionated towards a specific issue outlined in this [Quasar Forum Entry](https://groups.google.com/forum/#!topic/quasar-pulsar-user/IDbUt7MptcY). The original intent was to better integrate with the Arjuna transaction manager. A similar [discussion](https://developer.jboss.org/thread/269784) with JBoss devs.

- The fiber itself does not need to be exposed.
- The park callback should be executed in the same thread _before_ the thread locals are cleared.
- The unpark callback should be executed in the same thread that will execute the resuming fiber _after_ the thread locals are set.

**Comment by pron** [comment]
Hi. I have two reservations regarding this suggestion:

1. What is the API? How and when are the callbacks registered? I can think of two ways: either by subclassing `Fiber`, which would require special fiber classes to be used with certain libraries (and when using more than one such library, the callbacks won't compose), or by allowing dynamic registration of such callback, but in this case there will need to be special calls in the client code to do so, so the code using the library on fibers would look different from code using it in threads. This can be mitigated by wrapping the library API to register the callback whenever any of its methods is called.

2. More importantly, this isn't really a solution. It is a code-smell for any code to rely on thread identity (the correct way to do this is to rely on thread locals), and if some third-party code *does* rely on thread identity, there is usually nothing we can do. That a callback may work around this issue for a particular library is no more than a coincidence. Should we add an API that just happens to work around an issue that is in general difficult for some specific libraries (that may be rare)?

**Comment by circlespainter** [comment]
As suggested in the forum discussion I'd change the title to something broader like "Integrating frameworks that rely on thread identity".

Would a pinned fiber scheduler avoid the general problem?

**Comment by nickman** [comment]
I am a bit out on a limb on this since I don't have that much experience (about 52 hours) with the API, and no in-depth knowledge of the instrumentation, but this is based on the work I have done so far and the issue I am running into.

>>> What is the API ?

For me personally, I think these two methods in Fiber would be sufficient:

`protected void onPrePark();`

The method should be called by the fiber executing thread before threadlocals are cleared.

`protected void onPreResume();`

The method should be called by the thread that will be continuing the fiber after threadlocals are restored.

>>> It is a code-smell 

It does not smell to me, but I do defer to your expertise. I would add though, that this is not an uncommon pattern in a lot of products/libraries/frameworks, but is especially difficult to work around in transaction managers.  

>>>  Should we add an API that just happens to work around an issue that is in general difficult for some specific libraries (that may be rare)?

I hope so. 

I put this together in case it is helpful.

![image](https://cloud.githubusercontent.com/assets/597608/15295783/29f23884-1b60-11e6-9a29-98546a5aaed9.png)

**Comment by nickman** [comment]
Thanks for changing the title. 
I don't think a pinned fiber scheduler will resolve this (but I am going to try it). On first blush, I think it limits the number of running transactions to one, since upstream, the transaction manager only sees one thread.

**Comment by circlespainter** [comment]
@nickman Quasar at present doesn't provide a pinned multi-threaded fiber scheduler. What you could try is a single-threaded scheduler (or a set of them, performing fiber creation by round-robin on them) which is indeed a trivial pinned scheduler.

If the framework not only checks limits access to a transaction by a given thread, but doesn't even allow a single thread to handle more than one transaction concurrently then any scheduler will limit the number of concurrent transactions to its maximum number of worker threads. Furthermore, new fibers will fail creating new transactions if they get allocated to the same worker thread where a fiber has already started one. In this case I think there's really nothing that can help but to give up and use threads rather than fibers.

On the other hand this worst-case scenario might not be terrible: which kind of transactions are we talking about? If those are I/O-bound transactions (e.g. DB) then I doubt starting lots and lots of them concurrently will benefit the application performance, or possibly won't even be manageable by the I/O subsystem.

Finally, in the specific case of Arjuna, we're awaiting to understand if the warning printed by the is harmless (and it might well be).

**Comment by nickman** [comment]
Agreed. I understand the concept, I just can't be sure how it will work in
practice, which is on me.

Just to clarify, I think most JTA/JTS transaction managers are not bothered
by multiple threads executing within a transaction, just not concurrently
and in order to keep their housekeeping in order, they require a
suspend/resume when switching threads.

Thank-you again for your time on this.

//Nicholas

**Comment by pron** [comment]
@nickman Problem is, your proposed solution doesn't solve the problem, as you won't be able to use the callbacks. Your solution requires subclassing `Fiber`, which has two problems:

1. It will only work in certain kinds of fibers, which you may not always be able to control.
2. It does not compose. What if two different libraries require use of those callbacks? You'll have two different subclasses.

The best solution -- and the correct one, especially in light of various asynchronous frameworks -- is for the transaction library not to rely on thread identity. Instead, it should accept a function that returns a current execution context (`Object currentExecutionContext()`). The default implementation can return `Thread.currentThread().getId()`, and the Quasar-compatible one would be `Strand.currentStrand().getId()`. When used in asynchronous libraries (such as RxJava), a different type of execution context would be returned. Instead of comparing the current thread, the library should compare the current execution context.

The alternative would be a callback that can be dynamically attached to a fiber (rather than by subclassing), and some mechanism by which any relevant fiber would automatically have this callback registered. I think such a solution is both too excessive and too specific (as it only helps some very specific libraries, that happen to have suspension methods and rely on thread locals in a very specific way).

**Comment by nickman** [comment]
pron;

This was my reasoning:

For item #1, I created a new FiberFactory (*TXFiberFactory*) that creates
*TXFiberThread*s. When I spawn my actor instances, I pass an instance of
TXFiberFactory, so the actor's fiber is the type I want. I realize this is
not applicable to a variety of other scenarios.

For item #2, I had not considered that, but if I had that issue, [off the
top of my head] I would create an intermediate callback handler where all
my other callback listeners would register. I mean, there's a number of
ways to skin that cat, but not so many without the callback.

I like your second (alternative) solution, but it does not sound like you
do.
Your first solution sounds good too, but it shifts the burden into code I
cannot touch, namely third party libraries that are not condusive to
alteration. I will look further into the Arjuna framework and see if there
is way to implement the functionality you outlined, but I suspect it will
require some surgery.

The JBoss folks suggested looking at Arjuna's ThreadData
<https://github.com/jbosstm/narayana/blob/master/ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/thread/ThreadActionData.java>
class to save and restore the context but as far as I can tell, as soon as
a library makes use of ThreadLocals to switch contexts, it's a non-starter
because by the time a fiber is parked, all those ThreadLocals have been
made inaccessible to the very thread that needs to access them. But I will
keep digging into it.... this itch in the back of my head is making me
thing there's something there.

Cheers.

//Nicholas
